iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/grammar/DockerLexicalConstant.java

```diff
@@ -30,11 +30,42 @@ public class DockerLexicalConstant {
   public static final String STRING_LITERAL_WITH_QUOTES = "\"(?:[^\"\\\\]*+(?:\\\\[\\s\\S])?+)*+\"";
   public static final String STRING_LITERAL_WITHOUT_QUOTES = "(?:(?!" + LINE_BREAK+ ")[^\\s])++";
   public static final String STRING_LITERAL_WITHOUT_QUOTES_NO_EQUALS = "[^\\s=]++";
-  public static final String STRING_LITERAL = "(?:(?:" + STRING_LITERAL_WITH_QUOTES + ")|(?:" + STRING_LITERAL_WITHOUT_QUOTES + "))+";
+  public static final String STRING_LITERAL_OLD = "(?:(?:" + STRING_LITERAL_WITH_QUOTES + ")|(?:" + STRING_LITERAL_WITHOUT_QUOTES + "))+";
   public static final String KEY_IN_KEY_VALUE_PAIR_IN_EQUALS_SYNTAX = "(?:" + STRING_LITERAL_WITH_QUOTES + ")|(?:" + STRING_LITERAL_WITHOUT_QUOTES_NO_EQUALS + ")";
   public static final String STRING_UNTIL_EOL = ".+";
   public static final String EQUALS_OPERATOR = "=";
 
+  /**
+   * IDENTIFIERS
+   */
+
+  /**
+   * LITERAL
+   */
+  private static final String IDENTIFIER_START = "[a-zA-Z_\\x7f-\\xff]";
+
+  /**
+   * '$' sign is allowed in double quoted string and heredoc only when it does not conflict with the
+   * encapsulated variable expression, i.e when it not followed with '{' or a starting identifier character.
+   */
+  private static final String PERMITTED_EMBEDDED_DOLAR = "(?:\\$(?!\\{|" + IDENTIFIER_START + "))";
+
+  private static final String NON_SPECIAL_CHARACTERS = "(?:[^\"\\\\$\\{])";
+
+  // TODO: Handle custom escaping for escaped characters in strings
+  private static final String ESCAPED_CHARACTER_OR_STANDALONE_BACKSLASH = "(?:\\\\[\\s\\S]?)";
```

Comment:
Note sure what you mean with this comment.
Also, the directive `# escape='` is not just about defining the multi-line operator : this is the global escape operator.
I guess it can be addressed in another PR, also maybe this is what your comment is about ?